File: src/main/java/com/qtech/bubbles/entity/types/EntityType.java
package com.qtech.bubbles.entity.types;

import com.qtech.bubbles.common.gametype.AbstractGameType;
import com.qtech.bubbles.entity.Entity;
import com.qtech.bubbles.registry.RegistryEntry;
import org.bson.BsonDocument;

import java.util.Objects;

public class EntityType<T extends Entity> extends RegistryEntry {
    private final EntityFactory<T> entityFactory;

    public EntityType(EntityFactory<T> entityFactory) {
        this.entityFactory = entityFactory;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        EntityType<?> that = (EntityType<?>) o;
        return Objects.equals(getRegistryName(), that.getRegistryName());
    }

    public T create(AbstractGameType gameType) {
        return entityFactory.create(gameType);
    }

    @Override
    public int hashCode() {
        return Objects.hash(getRegistryName());
    }

    public T create(AbstractGameType gameType, BsonDocument document) {
        T t = entityFactory.create(gameType);
        t.setState(document);
        return t;
    }
}

File: src/main/kotlin/com/qtech/bubbles/entity/types/EntityType.kt
package com.qtech.bubbles.entity.types

import com.qtech.bubbles.common.RegistryEntry
import com.qtech.bubbles.entity.Entity
import com.qtech.bubbles.common.gametype.AbstractGameMode
import net.querz.nbt.tag.CompoundTag
import java.util.*

open class EntityType<T : Entity>(private val entityFactory: EntityFactory<T>) : RegistryEntry() {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as EntityType<*>
        return registryName == that.registryName
    }

    fun create(gameMode: AbstractGameMode): T {
        return entityFactory.create(gameMode)
    }

    override fun hashCode(): Int {
        return Objects.hash(registryName)
    }

    fun create(gameMode: AbstractGameMode, nbt: CompoundTag): T {
        val t = entityFactory.create(gameMode)
        t.setState(nbt)
        return t
    }
}